omf/blockmodel/model.py

```diff
@@ -0,0 +1,145 @@
+"""blockmodel/models.py: block model element and grids."""
+import numpy as np
+import properties
+
+from ..base import BaseModel, ProjectElement
+from .index import ijk_to_index, index_to_ijk
+from .subblock_check import subblock_check
+from .subblocks import FreeformSubblocks, RegularSubblocks
+
+__all__ = ["BlockModel", "RegularGrid", "TensorGrid"]
+
+
+class RegularGrid(BaseModel):
+    """Describes a regular grid of blocks with equal sizes."""
+
+    schema = "org.omf.v2.elements.blockmodel.grid.regular"
+
+    block_count = properties.Array("Number of blocks in each of the u, v, and w directions.", dtype=int, shape=(3,))
+    block_size = properties.Vector3("Size of blocks in the u, v, and w directions.", default=lambda: (1.0, 1.0, 1.0))
+
+    @properties.validator("block_count")
+    def _validate_block_count(self, change):
+        for item in change["value"]:
+            if item < 1:
+                raise properties.ValidationError("block counts must be >= 1", prop=change["name"], instance=self)
+
+    @properties.validator("block_size")
+    def _validate_block_size(self, change):
+        for item in change["value"]:
+            if item <= 0.0:
+                raise properties.ValidationError("block sizes must be > 0.0", prop=change["name"], instance=self)
+
+
+class TensorGrid(BaseModel):
+    """Describes a grid with varied spacing in each direction."""
+
+    schema = "org.omf.v2.elements.blockmodel.grid.tensor"
+
+    tensor_u = properties.Array("Tensor cell widths, u-direction", dtype=float, shape=("*",))
+    tensor_v = properties.Array("Tensor cell widths, v-direction", dtype=float, shape=("*",))
+    tensor_w = properties.Array("Tensor cell widths, w-direction", dtype=float, shape=("*",))
+
+    @properties.validator("tensor_u")
+    @properties.validator("tensor_v")
+    @properties.validator("tensor_w")
+    def _validate_tensor(self, change):
+        for item in change["value"]:
+            if item <= 0.0:
+                raise properties.ValidationError("tensor sizes must be > 0.0", prop=change["name"], instance=self)
+
+    @property
+    def block_count(self):
+        """The block count is derived from the tensors here."""
+        counts = []
+        for tensor in self.tensor_u, self.tensor_v, self.tensor_w:
+            if tensor is None:
+                return None
+            counts.append(len(tensor))
+        return np.array(counts, dtype=int)
+
+
+class BlockModel(ProjectElement):
+    """A block model, details are in the definition and sub-blocks attributes.
+
+    Sub-blocking is stored in the `subblocks` attribute. Use :code:`None` if there are no
```

Comment:
Is the sub-blocking description needed given that is in the subblocks attribute?

Reply:
Ah, I should have removed the other long docs. The `properties` module screws up docs beyond one line on attributes.